runtime/utils/common/src/aws/smithy/kotlin/runtime/util/net/Host.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.util.net
+
+/**
+ * A [Host] represents a parsed internet host. This may be an internet address (IPv4, IPv6) or a domain name.
+ */
+public sealed class Host {
```

Comment:
I'm not 100% certain I agree an ADT is better here but I can understand the desire. Curious if there was discussion on this and the tradeoffs?

Reply:
Yeah, so basically, the semantics I originally had was you'd have a string and then you could call `isIpv4` / `isIpv6` to figure out what it was.

The ADT was Ian's suggestion further up, which I went with because I'm a strong proponent of assigning explicitness to things where we can (as opposed to having a contextless string that you have to sniff the format to determine what it is).

The only _real_ downside I see is it's a bit more effort to work with the value (in that you have to wrap/unwrap from/to string).

Reply:
Ok. Yeah I don't have a super strong opinion either way. I suppose since we have a constructor for a `Url` from string directly and `Host.parse()` that those should be sufficient.